### ISSUE TITLE
GroupedLinear experiences significant performance degradation under bfloat16 when m_split is uneven.

### ISSUE THREAD

**Issue by littsk** [NVIDIA/TransformerEngine#2151]
**Describe the bug**

GroupedLinear experiences significant performance degradation under bfloat16 when m_split is uneven.

**Steps/Code to reproduce bug**

`import torch
from transformer_engine.pytorch.cpp_extensions.gemm import general_grouped_gemm
from transformer_engine.pytorch.module.base import get_multi_stream_cublas_workspace

tokens = 4096
in_features = 4096
out_features = 8192
num_experts = 4

inp = torch.arange(0, tokens, device="cuda", dtype=torch.bfloat16).repeat_interleave(in_features).reshape(-1, in_features)
weight = torch.randn((num_experts, out_features, in_features), device="cuda", dtype=torch.bfloat16)
weight_list = torch.unbind(weight, dim=0)
m_splits_balance = [1024, 1024, 1024, 1024]
m_splits_unbalance = [1042, 1024, 1018, 1012]

out = torch.empty((tokens, out_features), device="cuda", dtype=torch.float32)

torch.cuda.profiler.start()

# balanced case
for i in range(10):
    inp_list = torch.split(inp, m_splits_balance, dim=0)
    
    with torch.cuda.nvtx.range(f"balanced_{i}"):
        general_grouped_gemm(
            inp_list,
            weight_list,
            out=[out],
            out_dtype=torch.float32,
            workspaces=get_multi_stream_cublas_workspace(),
            m_splits=m_splits_balance,
            single_output=True,
        )

# unbalanced case
for i in range(10):
    inp_list = torch.split(inp, m_splits_unbalance, dim=0)
    
    with torch.cuda.nvtx.range(f"unbalanced_{i}"):
        general_grouped_gemm(
            inp_list,
            weight_list,
            out=[out],
            out_dtype=torch.float32,
            workspaces=get_multi_stream_cublas_workspace(),
            m_splits=m_splits_unbalance,
            single_output=True,
        )

torch.cuda.profiler.stop()`

**Behavior**

<img width="2094" height="958" alt="Image" src="https://github.com/user-attachments/assets/d38893a5-277a-4b7c-a4d2-e0748796fe7f" />
In unbalanced scenarios, performance drops significantly, and tensor core activity remains very low.

**Environment details**
- NGC container: nvcr.io/nvidia/pytorch:25.05-py3
- OS 24.04.2 LTS
- PyTorch 2.8.0a0+5228986c39.nv25.05
- Python 3.12.3
- Transformer Engine 2.3.0+5de3e14
- CUDA V12.9.41
- CUDNN 9.10.1

**Device details**
- On H100

**Comment by yaox12** [comment]
This is because you're using M, which is not 16B aligned, as the leading dimension. So cuBLAS cannot use a TMA-based kernel, and has to fall back to a naive implementation.

In TE, we always calculate `weight @ input` to let K become the leading dimension [[code](https://github.com/NVIDIA/TransformerEngine/blob/11e9d669ae827b13dce309fefa79f8938da34352/transformer_engine/pytorch/module/grouped_linear.py#L171-L173)] to obtain a good perf.

I modified your script to
```python
import torch
from transformer_engine.pytorch.cpp_extensions.gemm import general_grouped_gemm
from transformer_engine.pytorch.module.base import get_multi_stream_cublas_workspace

tokens = 4096
in_features = 4096
out_features = 8192
num_experts = 4


inp = torch.randn((tokens, in_features), device="cuda", dtype=torch.bfloat16)
weight = torch.randn((num_experts, out_features, in_features), device="cuda", dtype=torch.bfloat16)
weight_list = torch.unbind(weight, dim=0)
m_splits_balance = [1024, 1024, 1024, 1024]
m_splits_unbalance = [1042, 1024, 1018, 1012]

out = torch.empty((tokens, out_features), device="cuda", dtype=torch.float32)

torch.cuda.profiler.start()

# balanced case
for i in range(10):
    inp_list = torch.split(inp, m_splits_balance, dim=0)
    
    with torch.cuda.nvtx.range(f"balanced_{i}"):
        general_grouped_gemm(
            weight_list,
            inp_list,
            out=[out],
            out_dtype=torch.float32,
            workspaces=get_multi_stream_cublas_workspace(),
            m_splits=m_splits_balance,
            single_output=True,
        )

# unbalanced case
for i in range(10):
    inp_list = torch.split(inp, m_splits_unbalance, dim=0)
    
    with torch.cuda.nvtx.range(f"unbalanced_{i}"):
        general_grouped_gemm(
            weight_list,
            inp_list,
            out=[out],
            out_dtype=torch.float32,
            workspaces=get_multi_stream_cublas_workspace(),
            m_splits=m_splits_unbalance,
            single_output=True,
        )

torch.cuda.profiler.stop()
```
There is almost no perf gap between even and uneven m_splits.

**Comment by littsk** [comment]
This method has effectively solved my problem. Thank you for your feedback! 😁
